crates/sui-indexer-alt-jsonrpc/src/api/objects/filter.rs

```diff
@@ -64,50 +82,356 @@ type Cursor = BcsCursor<ObjectCursor>;
+    fn matches(&self, info: &StoredObjInfo) -> bool {
+        let (Some(package), Some(module), Some(name), Some(instantiation)) =
+            (&info.package, &info.module, &info.name, &info.instantiation)
+        else {
+            // If any of these fields are `None`, the record is for a deleted object which cannot
+            // match any filters.
+            return false;
+        };
+
+        use RawFilter as R;
+        let (package_filter, module_filter, name_filter, instantiation_filter) = match self {
+            R::MatchNone(raw_filters) => return !raw_filters.iter().any(|f| f.matches(info)),
+
+            R::Type {
+                package,
+                module,
+                name,
+                instantiation,
+            } => (
+                package,
+                module.as_ref(),
+                name.as_ref(),
+                instantiation.as_ref(),
+            ),
+        };
+
+        if package_filter != package {
+            return false;
+        }
+
+        if module_filter.is_some_and(|m| m != module) {
+            return false;
+        }
+
+        if name_filter.is_some_and(|n| n != name) {
+            return false;
+        }
+
+        if instantiation_filter.is_some_and(|i| i != instantiation) {
+            return false;
+        }
+
+        true
     }
 }
 
 /// Fetch ObjectIDs for a page of objects owned by `owner` that satisfy the given `filter` and
-/// pagination parameters. Returns the digests and a cursor point to the last result (if there are
+/// pagination parameters. Returns the digests and a cursor pointing to the last result (if there are
 /// any results).
 pub(super) async fn owned_objects(
     ctx: &Context,
     owner: SuiAddress,
     filter: &Option<SuiObjectDataFilter>,
     cursor: Option<String>,
     limit: Option<usize>,
+) -> Result<ObjectIDs, RpcError<Error>> {
+    match filter {
+        Some(f) if f.is_compound() => {
+            by_sequential_scan(ctx, owner, &f.to_raw(ctx)?, cursor, limit).await
+        }
+        filter => by_type_indices(ctx, owner, filter, cursor, limit).await,
+    }
+}
+
+/// Fetch ObjectIDs for a page of objects owned by `owner` that satisfy the given compound
+/// `filter`. Works by repeatedly fetching pages of objects owned by the owner, filtering out only
+/// matching entries until the limit is met.
+async fn by_sequential_scan(
+    ctx: &Context,
+    owner: SuiAddress,
+    filter: &RawFilter,
+    cursor: Option<String>,
+    limit: Option<usize>,
+) -> Result<ObjectIDs, RpcError<Error>> {
+    let config = &ctx.config().objects;
+
+    let page: Page<Cursor> = Page::from_params(
+        config.default_page_size,
+        config.max_page_size,
+        cursor,
+        limit,
+        None,
+    )?;
+
+    // Initially, be optimistic and assume that all the results fetched will match.
+    let mut results = vec![];
+    let mut cursor = page.cursor.map(|c| c.0);
+    let mut fetch = page.limit + 1;
```

Comment:
Interesting. So we first are optimistic and only fetch the limit, then starting from the second iteration we become pessimistic and fetch scan limit.

Reply:
Yes -- originally, I was going to tune the size of the next page to over-fetch according to the number of rows we have fetched so far, over the number of rows we actually kept (i.e. based on an assumption that the ratio of rows discarded in the next page matches the ratio we have seen so far).

I decided to do something simpler initially (and add metrics to check how much we're fetching), because I found it tricky to convince myself that this kind of feedback loop couldn't get stuck in quite a bad state because of rounding errors, etc.